src/java/org/apache/cassandra/repair/RepairRunnable.java

```diff
@@ -131,53 +146,128 @@ protected void fireProgressEvent(ProgressEvent event)
-        final TraceState traceState;
-        final UUID parentSession = UUIDGen.getTimeUUID();
-        final String tag = "repair:" + cmd;
+        long durationMillis = System.currentTimeMillis() - creationTimeMillis;
+        if (msg == null) {
+            String duration = DurationFormatUtils.formatDurationWords(durationMillis,true, true);
+            msg = String.format("Repair command #%d finished in %s", cmd, duration);
+        }
 
-        final AtomicInteger progress = new AtomicInteger();
-        final int totalProgress = 4 + options.getRanges().size(); // get valid column families, calculate neighbors, validation, prepare for repair + number of ranges to repair
+        fireProgressEvent(new ProgressEvent(ProgressEventType.COMPLETE, progressCounter.get(), totalProgress, msg));
 
-        String[] columnFamilies = options.getColumnFamilies().toArray(new String[options.getColumnFamilies().size()]);
-        Iterable<ColumnFamilyStore> validColumnFamilies;
-        try
+        if (ActiveRepairService.instance.hasParentRepairSession(parentSession))
+            ActiveRepairService.instance.removeParentRepairSession(parentSession);
+        Context ctx = context;
+        if (options.isTraced() && ctx != null && ctx.traceState != null)
         {
-            validColumnFamilies = storageService.getValidColumnFamilies(false, false, keyspace, columnFamilies);
-            progress.incrementAndGet();
+            for (ProgressListener listener : listeners)
+                ctx.traceState.removeProgressListener(listener);
+            // Because DebuggableThreadPoolExecutor#afterExecute and this callback
+            // run in a nondeterministic order (within the same thread), the
+            // TraceState may have been nulled out at this point. The TraceState
+            // should be traceState, so just set it without bothering to check if it
+            // actually was nulled out.
+            Tracing.instance.set(ctx.traceState);
+            Tracing.traceRepair(msg);
+            Tracing.instance.stopSession();
+        }
+
+        Keyspace.open(keyspace).metric.repairTime.update(durationMillis, TimeUnit.MILLISECONDS);
+    }
+
+    public void run()
+    {
+        try {
+            // Why is this before setup/start when its publishing the start event?  For backwards compatability
+            // One of the first things we did before was publish this before validating, so publish early to keep
+            // that the same.
+            String message = String.format("Starting repair command #%d (%s), repairing keyspace %s with %s", cmd, parentSession, keyspace,
+                                           options);
+            logger.info(message);
+            Tracing.traceRepair(message);
+            fireProgressEvent(new ProgressEvent(ProgressEventType.START, 0, 100, message));
+
+            Pair<Context, String> setup = setup();
```

Comment:
Its not immediately obvious to the reader what the `Pair` represents. I would prefer if you had a nested static class with descriptive fields like you've done with the `Context`.

Reply:
I added Either...